### PR TITLE
OOP School Library: Set up Associations

### PR DESCRIPTION
**Pull request by anyars-encarta**
In this project:

- [ ] I created a `Classroom` class with the following:
   - `@label` instance variable was initialized in the constructor.
   - Setter and getter were set for `@label` using `attr_accessor.`

- [ ] I created a `has-many`/`belongs-to` relationship between the `Classroom` and `Student` classes. The following were implemented:
   - I created the `has-many` side (a classroom has many students).
   - I created the `belongs-to` side (a student belongs to a classroom).
   - When adding a student to a classroom it also sets the classroom for the student.
   - When setting the classroom for a student it also adds it to the classrooms' students.

- [ ] I created a `Book` class with the following:
   - `@title` and `@author` instance variables were initialized in the constructor.
   - Setters and getters for instance variables were set using `attr_accessor`.

- [ ] I created a `Rental` class with the following:
   - `@date` instance variable was initialized in the constructor.
   - Setter and getter for `@date` was set using `attr_accessor`.

- [ ] I created the `many-to-many` (also `has-many-through`) relationship between `Person` and `Book` using the intermediate `Rental` class. The following were implemented:
   - I created the `has-many` side of `Book` and `Rental` (a book has many rentals).
   - I created the `belongs-to` side of `Rental` and `Book` (a rental belongs to a book).
   - I created the `has-many` side of `Person` and `Rental` (a person has many rentals).
   - I created the `belongs-to` side of `Rental` and `Person` (a rental belongs to a person).
   - I modified the constructor of `Rental` so `Book` and `Person` are set in it.

### CODE REVIEW

File: book.rb
Comment (Efiamotu-1) [inline book.rb:1]: - [ ] In your `readme.md` file, in the license section, the link you prvided to your license file is currently broken and it is unprofessional to leave broken links in your readme file so please kindly fix it :+1: :sparkles:

File: student.rb
Comment (Efiamotu-1) [inline student.rb:1]: - [ ] You are currently missing method that serves as the 'belongs-to' side (a student belongs to a classroom). see the below line of code as a guide to adding this to your project :+1: 
- code snippet
```ruby
def classroom=(classroom)
      @classroom = classroom
      classroom.students.push(self) unless classroom.students.include?(self)
  end
```

File: book.rb [hunk 1]
```diff
@@ -0,0 +1,11 @@
+class Book
+  attr_accessor :title, :author
```
Comment (Efiamotu-1) [inline book.rb]: - [ ] You are missing the attr_accessor for :rentals, please kindly add this to your codebase :+1:

File: book.rb
Comment (Efiamotu-1) [inline book.rb:1]: - [ ] `*A method for adding rentals should be present.` see the code snippet below for guidance 
- code snippet
```
def add_rental(person, date)
      Rental.new(date, self, person)
  end
```

File: person.rb [hunk 2]
```diff
@@ -7,10 +7,11 @@ def initialize(age, name = 'Unknown', parent_permission: true)
     @age = age
     @name = name
     @parent_permission = parent_permission
+    @rentals = []
   end
 
   attr_accessor :name, :age
-  attr_reader :id
+  attr_reader :id, :rentals
```
Comment (Efiamotu-1) [inline person.rb]: - [ ] You are missing the setter for :rentals, please kindly add this to your codebase using the attr_accessor keyword :+1:

File: person.rb
Comment (Efiamotu-1) [inline person.rb:1]: - [ ] `*A method for adding rentals should be present.` see the code snippet below for guidance :+1: 
- code snippet
```ruby
def add_rental(book, date)
      Rental.new(date, book, self)
  end
```

**Review by Efiamotu-1 — CHANGES_REQUESTED** [review]
Hi @anyars-encarta , 

 

Good job so far! 👏. Your Code is Awesome! 💪

There are some **few** issues 🐛 that you still need to work 🔧 on though to go to the next project ⏭️ but you are almost there! 🏃
## Don't Worry 🙏

It's ok to have bugs in your code 🐛. Please, see this code review as a **chance** for you to **learn** practical coding skills and learn **new tips** that can improve your code quality 💯 and help you **get a better ** 💼.

_I am going to guide you on how to fix them down below. 🤓_

## Highlights :dart: 
- ✔️ No linter Errors
- ✔️ Descriptive README
- ✔️ Descriptive PR title and description
- The code is top-notch! 💯 and has a huge potential.

## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._


## Important Notes 📝

- Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
- Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

> **Note:** Please, don't Merge This Pull Request yet. Only Merge when it is Approved ✅.


## Thank you for your time 🙏 and call me if you need help or clarification ☎️

- This is My **GitHub**: [@Efiamotu-1](https://github.com/Efiamotu-1/)
- This is My **Twitter**: [@EFYAMOTU](https://twitter.com/EFYAMOTU)
- This is My **linkedIn**.[Habeeb Efiamotu Musa](https://www.linkedin.com/in/musa-habeeb/)
- This is My **Slack**: [Habeeb Efiamotu Musa](https://slack.com)

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: person.rb [hunk 3]
```diff
@@ -25,4 +26,8 @@ def of_age?
   def can_use_services?
     true if of_age? || @parent_permission
   end
+
+  def add_rental(book, date)
+    Rental.new(date, book, self)
+  end
```
Comment (zeff96) [inline person.rb]: - [ ] Hey there! 🙌 I noticed that you've marked the "add person rental" method as private. 🤔 However, it appears that this method should be made public, not private! 📢 Let's make the adjustment to ensure it's accessible to everyone who needs it. 🚀

File: person.rb [hunk 4]
```diff
@@ -7,9 +7,10 @@ def initialize(age, name = 'Unknown', parent_permission: true)
     @age = age
     @name = name
     @parent_permission = parent_permission
+    @rentals = []
   end
 
-  attr_accessor :name, :age
+  attr_accessor :name, :age, :rentals
```
Comment (zeff96) [inline person.rb]: - [  OPTIONAL ] Hi there! 🌟 It's widely considered a best practice to place Setter and Getter methods at the top of the class structure. 😊 Currently, they're located within the body of the class, which might make the code more organized if we move them up to the top. 🚀 This approach enhances code readability and makes it easier for fellow developers to understand and work with the class. 📜🤓

File: classroom.rb [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+
+  def add_student(student)
+    student&.classroom = self
+    @students << student
+  end
+
+  def classroom=(classroom)
+    @classroom = classroom
+    classroom.students.push(self) unless classroom.students.include?(self)
+  end
```
Comment (zeff96) [inline classroom.rb]: - [ ] Hey there! 🎓 To ensure your project aligns with the Keep It Simple (KISS) principle of Object-Oriented Programming, it's a great idea to move the method that represents the 'belongs-to' relationship (like a student belonging to a classroom) into the Student class within the student.rb file. 📦 This adjustment simplifies the structure, making it more intuitive and easy to manage, following OOP best practices. 🌟👨‍🎓🏫

**Review by zeff96 — CHANGES_REQUESTED** [review]
### Status: CHANGES REQUIRED ♻️
Hi @anyars-encarta , 

Great work on making the changes requested by a previous reviewer 👏 
You've done well implementing some of the requested changes, but there are still some which aren't addressed yet.

almost

 ## Highlights
✔️ Linters passing
✔️ Correct use of GitFlow

 ## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._


Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**


------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: classroom.rb (hunk 5 above)
Comment (anyars-encarta) [reply classroom.rb]: @zeff96 , When I add this method into the student.rb file. 
```
def classroom=(classroom)
    @classroom = classroom
    classroom.students.push(self) unless classroom.students.include?(self)
  end
```

I get a duplicate lint error:
`Lint/DuplicateMethods: Method `Student#classroom=` is defined at both student.rb:4 and student.rb:16.`

**Review by Gambit142 — APPROVED** [review]
Hi @anyars-encarta 👋,

## Status: Approved 🟢   

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉

### Highlights
- Setup linters correctly ✔️ 
- Set up the association between classes correctly ✔️ 
- Has `setters` and `getters` methods for classes ✔️ 

Cheers, and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.

#### Please tag me in your question so I can receive the notification. You can tag me by typing @Gambit142

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think the code review was unfair, request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**